packages/components/src/List/Toolbar/ColumnChooser/hooks/columnChooserManager.hook.js

```diff
@@ -0,0 +1,154 @@
+import { useState } from 'react';
+import cloneDeep from 'lodash/cloneDeep';
+import invariant from 'invariant';
+import flow from 'lodash/flow';
+import { compareOrder } from '../service';
+
+export function isValueCorrect(value, collectionLength) {
+	return Number.isInteger(value) && (value <= collectionLength || value < 0);
+}
+
+export function getOrderItem(order, index, length) {
+	if (index === length - 1) {
+		return order - 1;
+	}
+	return order + 1;
+}
+
+export function matchOrder(value) {
+	return function match({ order }) {
+		return order === value;
+	};
+}
+
+export function incrementColumnOrder(column, index) {
+	return { ...column, order: index + 1 };
+}
+
+export function organiseEditedColumns(collection) {
+	return collection.sort(compareOrder).map(incrementColumnOrder);
+}
+
+export function changeColumnAttribute(key) {
+	return function setAttribut(value, index) {
+		return function setColumn(collection) {
+			const newColumns = collection;
+			newColumns[index][key] = value;
+			return collection;
+		};
+	};
+}
+
+export function updateEditedColumns(editedColumns) {
+	return function updateState(state) {
+		return {
+			...state,
+			editedColumns,
+		};
+	};
+}
+
+const updateAttributeVisiblity = changeColumnAttribute('hidden');
+const updateAttributeOrder = changeColumnAttribute('order');
+
+export function useColumnChooserManager(columns, customSubmit) {
+	const [state, setState] = useState({
+		editedColumns: organiseEditedColumns(columns),
+		selectAll: false,
+	});
+
+	function getEditedColumnsLength() {
+		return state.editedColumns.length;
+	}
+
+	function getValueFormated(value) {
+		const checkedValue = isValueCorrect(parseInt(value, 10), getEditedColumnsLength());
+		invariant(checkedValue, `ColumnChooserManager, getValueFormated : Bad order number ${value}`);
+		return checkedValue;
+	}
+
+	function modifyOrderTwoItems(value, index) {
+		const indexColToReplace = state.editedColumns.findIndex(matchOrder(value));
+		const orderToReplace = getOrderItem(value, indexColToReplace, getEditedColumnsLength());
+		if (indexColToReplace > -1 && !state.editedColumns[indexColToReplace].locked) {
+			flow([
+				updateAttributeOrder(orderToReplace, indexColToReplace),
+				updateAttributeOrder(value, index),
+				organiseEditedColumns,
+				updateEditedColumns,
+				setState,
+			])(state.editedColumns);
+		}
+	}
+
+	function handlerChangeVisibility(index) {
+		return function changeVisiblity(event, value) {
+			flow([updateAttributeVisiblity(value, index), updateEditedColumns, setState])(
+				state.editedColumns,
+			);
+		};
+	}
+
+	function handlerInputTextOrder(index) {
+		return function inputOrder(event, value) {
+			let parseValue;
+			try {
+				parseValue = getValueFormated(value);
```

Comment:
The value is from `getValueFormated ()` that returns the result of `isValueCorrect ()`
The types are not clear because we expect `isValueCorrect ()` to be a predicate, so a boolean.

Reply:
Yep fixed, moved some code without thinking